File: app/src/main/java/com/frogobox/board/mvvm/tutorial/TutorialActivity.java
package com.frogobox.board.mvvm.tutorial;

import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.os.Build;
import android.os.Bundle;
import android.text.Html;
import android.view.LayoutInflater;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.core.content.ContextCompat;
import androidx.viewpager.widget.ViewPager;

import com.bumptech.glide.Glide;
import com.frogobox.board.R;
import com.frogobox.board.core.BaseActivity;
import com.frogobox.board.mvvm.main.MainActivity;

public class TutorialActivity extends BaseActivity {

    private static final String TAG = TutorialActivity.class.getSimpleName();
    private ViewPager viewPager;
    private LinearLayout dotsLayout;
    private Button btnSkip, btnNext;
    private TutorialController tutorialController;

    // layouts of all welcome sliders
    // add few more layouts if you want
    private int[] layouts = new int[]{
            R.layout.fragment_tutorial_slide,
            R.layout.fragment_tutorial_slide,
            R.layout.fragment_tutorial_slide,
            R.layout.fragment_tutorial_slide,
    };
    //  viewpager change listener
    ViewPager.OnPageChangeListener viewPagerPageChangeListener = new ViewPager.OnPageChangeListener() {

        @Override
        public void onPageSelected(int position) {


            addBottomDots(position);

            // changing the next button text 'NEXT' / 'GOT IT'
            if (position == layouts.length - 1) {
                // last page. make button text to GOT IT
                btnNext.setText(getString(R.string.okay));
                btnSkip.setVisibility(View.GONE);
            } else {
                // still pages are left
                btnNext.setText(getString(R.string.next));
                btnSkip.setVisibility(View.VISIBLE);
            }
        }

        @Override
        public void onPageScrolled(int arg0, float arg1, int arg2) {
        }

        @Override
        public void onPageScrollStateChanged(int arg0) {

        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_tutorial);

        // Making notification bar transparent
        if (Build.VERSION.SDK_INT >= 21) {
            getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_STABLE | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN);
        }

        tutorialController = new TutorialController(this);

        viewPager = findViewById(R.id.view_pager);
        dotsLayout = findViewById(R.id.layoutDots);
        btnSkip = findViewById(R.id.btn_skip);
        btnNext = findViewById(R.id.btn_next);


        // adding bottom dots
        addBottomDots(0);

        // making notification bar transparent
        changeStatusBarColor();

        LayoutInflater layoutInflater = (LayoutInflater) getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        TutorialPager tutorialPager = new TutorialPager(layoutInflater, layouts);
        viewPager.setAdapter(tutorialPager);
        viewPager.addOnPageChangeListener(viewPagerPageChangeListener);

        try {
            ImageView imageView = findViewById(R.id.iv_tutorial);
            Glide.with(TutorialActivity.this).load(R.drawable.ic_launcher).into(imageView);//.into(imageView);//@mipmap/ic_splash).into(imageView);
        } catch (NullPointerException ne) {

        }

        btnSkip.setOnClickListener(v -> launchHomeScreen());

        btnNext.setOnClickListener(v -> {
            // checking for last page
            // if last page home screen will be launched
            int current = getItem(+1);
            if (current < layouts.length) {
                // move to next screen
                viewPager.setCurrentItem(current);
            } else {
                launchHomeScreen();
            }
        });
    }

    private void addBottomDots(int currentPage) {
        TextView[] dots = new TextView[layouts.length];

        int activeColor = ContextCompat.getColor(this, R.color.dot_light_screen);
        int inactiveColor = ContextCompat.getColor(this, R.color.dot_dark_screen);

        dotsLayout.removeAllViews();
        for (int i = 0; i < dots.length; i++) {
            dots[i] = new TextView(this);
            dots[i].setText(Html.fromHtml("&#8226;"));
            dots[i].setTextSize(35);
            dots[i].setTextColor(inactiveColor);
            dotsLayout.addView(dots[i]);
        }

        if (dots.length > 0)
            dots[currentPage].setTextColor(activeColor);
    }

    private int getItem(int i) {
        return viewPager.getCurrentItem() + i;
    }

    private void launchHomeScreen() {
        if (tutorialController.isFirstTimeLaunch()) {
            Intent intent = new Intent(TutorialActivity.this, MainActivity.class);
            intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
            tutorialController.setFirstTimeLaunch(false);
            startActivity(intent);
        }
        finish();
    }

    /**
     * Making notification bar transparent
     */
    private void changeStatusBarColor() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            Window window = getWindow();
            window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
            window.setStatusBarColor(Color.TRANSPARENT);
        }
    }

}

File: app/src/main/java/com/frogobox/board/mvvm/tutorial/TutorialActivity.kt
package com.frogobox.board.mvvm.tutorial

import com.frogobox.board.R
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import android.os.Bundle
import android.view.LayoutInflater
import android.widget.TextView
import androidx.core.content.ContextCompat
import android.text.Html
import android.content.Intent
import android.graphics.Color
import android.view.View
import android.view.WindowManager
import com.frogobox.board.core.BaseBindingActivity
import com.frogobox.board.databinding.ActivityTutorialBinding
import com.frogobox.board.mvvm.main.MainActivity
import kotlinx.android.synthetic.main.fragment_tutorial_slide.*

class TutorialActivity : BaseBindingActivity<ActivityTutorialBinding>() {

    private val tutorialController by lazy {
        TutorialController(this)
    }

    private val layouts = intArrayOf(
        R.layout.fragment_tutorial_slide,
        R.layout.fragment_tutorial_slide,
        R.layout.fragment_tutorial_slide,
        R.layout.fragment_tutorial_slide
    )

    override fun setupViewBinding(): ActivityTutorialBinding {
        return ActivityTutorialBinding.inflate(layoutInflater)
    }

    override fun setupViewModel() {
    }

    override fun setupUI(savedInstanceState: Bundle?) {
        binding.apply {
            changeStatusBarColor()
            addBottomDots(0)

            setupViewPager()

            btnSkip.setOnClickListener {
                launchHomeScreen()
            }

            btnNext.setOnClickListener {
                val current = binding.viewPager.currentItem + 1
                if (current < layouts.size) {
                    viewPager.currentItem = current
                } else {
                    launchHomeScreen()
                }
            }
        }
    }

    private fun setupViewPager() {
        binding.apply {

            val layoutInflater = getSystemService(LAYOUT_INFLATER_SERVICE) as LayoutInflater
            val tutorialPager = TutorialPager(layoutInflater, layouts)

            viewPager.adapter = tutorialPager
            viewPager.addOnPageChangeListener(object : OnPageChangeListener {
                override fun onPageSelected(position: Int) {
                    addBottomDots(position)
                    // changing the next button text 'NEXT' / 'GOT IT'
                    if (position == layouts.size - 1) {
                        // last page. make button text to GOT IT
                        btnNext.text = getString(R.string.okay)
                        btnSkip.visibility = View.GONE
                    } else {
                        // still pages are left
                        btnNext.text = getString(R.string.next)
                        btnSkip!!.visibility = View.VISIBLE
                    }
                }

                override fun onPageScrolled(arg0: Int, arg1: Float, arg2: Int) {}
                override fun onPageScrollStateChanged(arg0: Int) {}
            })
        }
    }

    private fun addBottomDots(currentPage: Int) {
        binding.apply {
            val dots = arrayOfNulls<TextView>(layouts.size)
            val activeColor = ContextCompat.getColor(this@TutorialActivity, R.color.dot_light_screen)
            val inactiveColor = ContextCompat.getColor(this@TutorialActivity, R.color.dot_dark_screen)
            dotsLayout.removeAllViews()
            for (i in dots.indices) {
                dots[i] = TextView(this@TutorialActivity)
                dots[i]!!.text = Html.fromHtml("&#8226;")
                dots[i]!!.textSize = 35f
                dots[i]!!.setTextColor(inactiveColor)
                dotsLayout.addView(dots[i])
            }
            if (dots.isNotEmpty()) dots[currentPage]!!.setTextColor(activeColor)
        }
    }

    private fun launchHomeScreen() {
        if (tutorialController.isFirstTimeLaunch) {
            val intent = Intent(this@TutorialActivity, MainActivity::class.java)
            intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP
            tutorialController.isFirstTimeLaunch = false
            startActivity(intent)
        }
        finish()
    }

    private fun changeStatusBarColor() {
        val windowS = window
        windowS.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
        windowS.statusBarColor = Color.TRANSPARENT
        windowS.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_STABLE or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
    }

}